include/libvfio-user.h

```diff
@@ -268,33 +246,106 @@ typedef enum {
 } vfu_pci_type_t;
 
 /**
- * Setup PCI configuration space header data. This function must be called only
- * once per libvfio-user context.
+ * Initialize the context for a PCI device. Returns an initialized buffer for a
+ * type 0 PCI configuration space of the needed size.
+ *
+ * This buffer consists of the standard 64-byte PCI header and the remaining
+ * space depending on @pci_type, either PCI_CFG_SPACE_SIZE or
+ * PCI_CFG_SPACE_EXP_SIZE.
  *
  * @vfu_ctx: the libvfio-user context
- * @id: Device and vendor ID
- * @ss: Subsystem vendor and device ID
- * @cc: Class code
  * @pci_type: PCI type (convention PCI, PCI-X mode 1, PCI-X mode2, PCI-Express)
  * @revision: PCI/PCI-X/PCIe revision
+ */
+vfu_pci_hdr_t *
+vfu_pci_init(vfu_ctx_t *vfu_ctx, vfu_pci_type_t pci_type, int revision UNUSED);
+
+/*
+ * FIXME
+ */
+void
+vfu_pci_set_id(vfu_ctx_t *vfu_ctx, uint16_t vid, uint16_t did,
+               uint16_t ssid, uint16_t ssvid);
+
+/*
+ * FIXME
+ */
+void
+vfu_pci_set_class(vfu_ctx_t *vfu_ctx, uint8_t base, uint8_t sub, uint8_t pi);
+
+// FIXME: what else do we need? multi-function flag in "header type" ? rid?
+
+/**
+ * Add a PCI capability to PCI config space.
+ *
+ * Certain standard capabilities are handled entirely within the library:
+ *
+ * PCI_CAP_ID_EXP (pxcap)
+ * PCI_CAP_ID_MSIX (msixcap)
+ * PCI_CAP_ID_PM (pmcap)
+ *
+ * However, they must still be explicitly initialized and added here.
  *
- * @returns 0 on success, -1 on failure and sets errno.
+ * FIXME: are we really making the user init these too? why? Is there a better
+ * API?
+ *
+ * The contents of @data is copied in. It must start with either a struct
+ * pci_cap or a struct pci_extcap, with the ID field set (the 'next' field is
+ * ignored).  If the capability is not one of the handled ones above (e.g.
+ * PCI_CAP_ID_VNDR), the library will allow all reads and disallow all writes
+ * for the capability from the client.
```

Comment:
I don't know enough about capabilities to tell whether it's OK to allow reads to go unnoticed by the device, what if some have register semantics? I think it safer to simply state that currently we only handle a small set of capabilities and that we can accept contributions for more capabilities.

Reply:
We have to do *something* if we're not doing a cb.

Reply:
> We have to do _something_ if we're not doing a cb.

I think we should not allow capabilities that are not in the list of capabilities we support then `vfu_pci_add_capability` should fail with `ENOTSUP`. This way we won't have to do anything.

Reply:
that won't help with vendor caps: we know we have to support them, but we can't implement them.

Reply:
For now we can allow the host to directly read/write to a vendor-specific capability. If the device needs to handle accesses, then we can introduce a function that sets a callback for a vendor-specific capability. I implemented [support for the vendor-specific capability](https://github.com/nutanix/libvfio-user/pull/177/commits/4da7d7dd441b0f13c0ba423e74f3bfe0265a9e4e) and IMO this is the best way to solve the problem for now.

Reply:
I'm not sure about the "write" part of that, hence my suggestion above. We could add a "READONLY" flag maybe?